crypto/bls12381/bls12381.go

```diff
@@ -0,0 +1,189 @@
+	blsPrivateKey, err := bls.PrivateKeyFromBytes(privKey, true)
+	if err != nil {
+		return nil, err
+	}
+	insecureSignature := blsPrivateKey.SignInsecure(msg)
+	return insecureSignature.Serialize(), nil
+}
+
+// PubKey gets the corresponding public key from the private key.
+//
+// Panics if the private key is not initialized.
+func (privKey PrivKey) PubKey() crypto.PubKey {
+	// set modOrder flag to true so that too big random bytes will wrap around and be a valid key
+	blsPrivateKey, err := bls.PrivateKeyFromBytes(privKey, true)
+	if err != nil {
+		// should probably change method sign to return an error but since
+		// that's not available just panic...
+		panic("bad key")
+	}
+	publicKeyBytes := blsPrivateKey.PublicKey().Serialize()
+	return PubKey(publicKeyBytes)
+}
+
+// Equals - you probably don't need to use this.
+// Runs in constant time based on length of the keys.
+func (privKey PrivKey) Equals(other crypto.PrivKey) bool {
+	if otherBLS, ok := other.(PrivKey); ok {
+		return subtle.ConstantTimeCompare(privKey[:], otherBLS[:]) == 1
+	}
+
+	return false
+}
+
+func (privKey PrivKey) TypeIdentifier() string {
+	return keyType
+}
+
+func (privKey PrivKey) Type() crypto.KeyType {
+	return crypto.BLS12381
+}
+
+// GenPrivKey generates a new bls12381 private key.
+// It uses OS randomness in conjunction with the current global random seed
+// in tendermint/libs/common to generate the private key.
+func GenPrivKey() PrivKey {
+	return genPrivKey(crypto.CReader())
+}
+
+// genPrivKey generates a new bls12381 private key using the provided reader.
+func genPrivKey(rand io.Reader) PrivKey {
+	seed := make([]byte, SeedSize)
+
+	_, err := io.ReadFull(rand, seed)
+	if err != nil {
+		panic(err)
+	}
+
+	return PrivKey(bls.PrivateKeyFromSeed(seed).Serialize())
+}
+
+// GenPrivKeyFromSecret hashes the secret with SHA2, and uses
+// that 32 byte output to create the private key.
+// NOTE: secret should be the output of a KDF like bcrypt,
+// if it's derived from user input.
+func GenPrivKeyFromSecret(secret []byte) PrivKey {
+	seed := crypto.Sha256(secret) // Not Ripemd160 because we want 32 bytes.
+
+	return PrivKey(bls.PrivateKeyFromSeed(seed).Serialize())
+}
+
+//-------------------------------------
+
+var _ crypto.PubKey = PubKey{}
+
+// PubKeyBLS12381 implements crypto.PubKey for the bls12381 signature scheme.
+type PubKey []byte
+
+// Address is the SHA256-20 of the raw pubkey bytes.
+func (pubKey PubKey) Address() crypto.Address {
+	if len(pubKey) != PubKeySize {
+		panic("pubkey is incorrect size")
+	}
+	return crypto.Address(tmhash.SumTruncated(pubKey))
+}
+
+// Bytes returns the PubKey byte format.
+func (pubKey PubKey) Bytes() []byte {
+	return []byte(pubKey)
+}
+
+func (pubKey PubKey) VerifyBytes(msg []byte, sig []byte) bool {
+	// make sure we use the same algorithm to sign
+	if len(sig) != SignatureSize {
+		return false
+	}
+	publicKey, err := bls.PublicKeyFromBytes(pubKey)
+	if err != nil {
+		// maybe log/panic?
```

Comment:
I do not recommend panicking in these methods.

Reply:
Got it